### DISCUSSION TITLE
How libfuse temporarily and elegantly pause receiving requests sent internally, and resume processing requests next time from kernal

### DISCUSSION THREAD

**Discussion by Janos-kz** [libfuse/libfuse#1072] [Q&A] [answered]
I have a CPP program that uses libfuse, and now I need this program to support hot upgrades. I need to pass the old session fd to the new process, but during this period, in order to ensure that I/O is not interrupted, I need to briefly stop libfuse from sending client requests, pass fd, and then restart fuse to resolve requests. I would like to ask how to modify libfuse. The version used is 3.16.2, and the reference is Juicefs' changes to the go fuse library https://github.com/juicedata/go-fuse/

**Comment by bsbernd** [comment]
I don't think that can be done from libfuse, it is fuse-client (kernel) that has to stop sending requests. I actually plan to implement that through notifications, just didn't come to it yet. I should find time for it within the next weeks.

**Reply by Janos-kz** [reply]
Okay, what I'm thinking is that the kernel can still send requests, but if they pile up and stop the loop at the libfuse layer from forwarding to the application layer, I'm not sure if it's feasible

**Comment by davidgaleano** [comment, marked as answer]
If you stop reading from the session fd then the requests will briefly accumulate on the kernel side until you resume reading from the rd again, the kernel will just fell a little hickup but it should be fine. How to stop reading from the fd would be up to your application and depends on how do you use libfuse.

**Reply by Janos-kz** [reply]
Yes, that's exactly what I intended. My program uses the low-level API mode of libfuse, which includes fuse_session-new, fuse_session-mount, and then fuse_session-loop_mt. So Currently, Is there a way to stop fuse_session-loop_mt with an interface and wait for the old program io to complete processing before running fuse_session-loop_t again on the old session id int new process to handle the accumulated kernel requests ?

**Reply by Janos-kz** [reply]
Is there a ready-made interface in libfuse library 3.16.2 that can achieve such a stop restore operation interface? If not, can I implement it myself? Is there a large amount of modification required ?

**Reply by ashleypittman** [reply]
It's easy enough to move the threading/looking functionality to a libfuse client, all you need to do is call `fuse_session_receive_buf`/`fuse_session_process_buf` in a loop across a number of threads.  I started by copying what `fuse_session_loop_mt` does and went from there.

That would get you the ability to pause reading from the fd if you wanted which might be enough for you to implement the rest yourself.

**Reply by bsbernd** [reply]
Please note that -over-io-uring is not using these functions, as one receives io-uring CQEs. And I hope that the kernel side is going to merged soon - I'm going to submit a non-RFC version this week. I need to test what signal io_uring_submit_and_wait() is acting on - maybe these can be woken up and put on hold.

**Reply by ashleypittman** [reply]
Ah, are you thinking that the io-uring stuff is going to be transparent to the libfuse user?  That would be awesome if so.

That said, the `loop_mt` function has some downsides, if io-uring is going to add to libfuse as a standard option we'll want to expand the threading interface to give more control.

**Reply by bsbernd** [reply]
Yes it is transparent. You _can_ benefit from it, as you have control over the request buffer. Just requires a check if fuse_req is from io-uring. I didn't create a PR for my uring branch yet, as the kernel design is not in stone yet and I just appended patches to that branch. The PR follows as soon as I get an ack for the design - maybe even this week.

**Reply by bsbernd** [reply]
Also, io-uring is not using loop_mt.  I hooks into fuse_lowlevel, but there is a header payload separation first. Eventually that will also be used for legacy /dev/fuse, as it avoids all the compat issues on struct size changes.

**Reply by bsbernd** [reply]
For DAOS you might want to use the polling interface and avoid additional ring threads at all. Actually still need to add that interface to example/passthrough_hp.

**Reply by Janos-kz** [reply]
> It's easy enough to move the threading/looking functionality to a libfuse client, all you need to do is call `fuse_session_receive_buf`/`fuse_session_process_buf` in a loop across a number of threads. I started by copying what `fuse_session_loop_mt` does and went from there.
> 
> That would get you the ability to pause reading from the fd if you wanted which might be enough for you to implement the rest yourself.

Got it, I found that I can stop the loop processing of fuse_session_loop_mt by using fuse_session_exit, and the fuse_session_loop_mt will wait internally for all requests to be processed and then I don't use fuse_session_unmount & fuse_session_destroy & fuse_session_reset, but transfer this fuse_session structure to another process, and run fuse_session_loop_mt again it will reprocess the previously blocked requests again. Can this truly stop passing IO from the libfuse layer to the user layer ?

**Comment by bsbernd** [comment]
I also have my own fuse-over-io-uring in my mind (https://lore.kernel.org/all/20241122-fuse-uring-for-6-10-rfc4-v6-0-28e6cdd0e914@ddn.com/) - you cannot simply send a signal to wake up these waiting IORING_OP_URING_CMDs. They only get woken up on a) kernel fuse sends a request b) the daemon gets killed.

**Comment by HereThereBeDragons** [comment]
i have to agree with bernd. i dont think there is any good way to do this right now.

we are also looking forward to having this natively available. right now we do a lot of hacky workarounds that work in most cases. but some rare corner cases like under heavy load, hick ups sadly still occure.

**Comment by bsbernd** [comment]
@HereThereBeDragons with plain /dev/fuse it would be indeed possible - threads that read on /dev/fuse would need to woken up with a signal and then libfuse would need to hold them back on reading /dev/fuse again until the pause is over. But I really wouldn't like to implement (or even accept) something that cause issues with other on going work, especially now that I'm hopefully close to get it merged.
